dump/dump.go

```diff
@@ -218,6 +227,89 @@ func getProxyConfiguration(ctx context.Context, group *errgroup.Group,
 	})
 }
 
+func addSharedPluginConfig(p *kong.Plugin, sharedPlugin utils.SharedPlugin) utils.SharedPlugin {
+	if p.Consumer != nil && p.Consumer.ID != nil {
+		sharedPlugin.Consumers = append(sharedPlugin.Consumers, *p.Consumer.ID)
+	}
+	if p.Service != nil && p.Service.ID != nil {
+		sharedPlugin.Services = append(sharedPlugin.Services, *p.Service.ID)
+	}
+	if p.Route != nil && p.Route.ID != nil {
+		sharedPlugin.Routes = append(sharedPlugin.Routes, *p.Route.ID)
+	}
+	return sharedPlugin
```

Comment:
The tracking here loses the relationship of a plugin with multiple entities.
Consider the following plugin:
```yaml
  plugins:
    name: rate-limiting
    service:
       id: svc-id
    consumer:
       id: consumer-id
    config:
      day: null
      fault_tolerant: true
      hide_client_headers: false
      hour: null
      limit_by: consumer
      minute: 10
      month: null
      policy: redis
      redis_database: 0
      redis_host: redis.common.svc
      redis_password: null
      redis_port: 6379
      redis_timeout: 2000
      second: null
      year: null
    enabled: true
    protocols:
    - http
    - https
```
In this case, this data structure would track the plugin as a plugin that is applied to the consumer, and the same plugin is applied to the service, in reality, that is not actually true, the plugin is present only once, on a combination of service and consumer.

The code that uses this data structure, later on, is structured in a way that one never runs into this bug currently. But that is a bit too risky to rely on. Please consider a data structure that captures the entire relationship explicitly.